### DISCUSSION TITLE
Nullable Reference Type information in Type class/type system

### DISCUSSION THREAD

**Discussion by skrysm** [dotnet/roslyn#50273] [Q&A] [answered]
First of, I'm not even sure whether this is the correct repo - or if dotnet/runtime or dotnet/csharplang would be better. So, sorry about that.

Now to my question: C# 8 introduced nullable reference types. I understand that they're implemented via attributes rather than in the type system itself. I'm guessing this is the reason why `typeof(string?)` doesn't compile (but fails with CS8639).

I would guess that while designing the nullable reference type feature there were discussions about how to bring this information into the type system (which the language designers decided against in the end).

Is this decision recorded anywhere - maybe even in a GitHub issue? Or is there even an open issue to bring this information into the type system in the future?

I've tried to find such an issue but couldn't find any.

**Comment by PathogenDavid** [comment, marked as answer]
The csharplang repo is probably a better choice for this question. If the decision is documented anywhere, it's likely in [the LDM meeting notes](https://github.com/dotnet/csharplang/tree/aed221e8c332cd430d41994149b4f5c64a37ba9c/meetings) or in https://github.com/dotnet/csharplang/issues/36.

The value of `typeof(T?)` is mentioned in https://github.com/dotnet/csharplang/issues/2201 but there are no notes associated with it in either LDM meeting associated with this issue. ([2019-02-13](https://github.com/dotnet/csharplang/blob/aed221e8c332cd430d41994149b4f5c64a37ba9c/meetings/2019/LDM-2019-02-13.md) and [2019-02-20](https://github.com/dotnet/csharplang/blob/aed221e8c332cd430d41994149b4f5c64a37ba9c/meetings/2019/LDM-2019-02-20.md)) There's also these comments: https://github.com/dotnet/csharplang/issues/36#issuecomment-285769923 https://github.com/dotnet/csharplang/issues/36#issuecomment-354468677

--------------------

If you want _my_ (very rambly) opinion ✨

It was probably never seriously considered to have `System.Type` record nullability. `System.Type` represents a type in the runtime, but nullability of reference types is a compile-time concept in C#. Additionally, anything that spans both the runtime and the compiler increases complexity of features significantly. Runtime changes are generally avoided at all costs.

Using attributes also allowed people to use preview compilers with old runtimes to test the feature and avoided changing the format of assemblies, meaning assemblies built with newer compilers can still be loaded in older runtimes.

Nullability is also arguably a characteristic of a local, field, property, function return value, parameter, etc. However if your goal is to get nullability information in reflection, it becomes problematic with more complex types like `IEnumerable<string?>` or `string?[][]?`. Which might make you want to associate nullability with `System.Type` anyway.

However, adding the concept of nullablility to `System.Type` opens a whole can of worms because a single `System.Type` instance can no longer correspond to a single runtime type. (Now all reference types effectively have two `System.Type` instances.) Does `typeof(string) == typeof(string?)`? (If yes, you break expectations. If no, you break reflection because...) Does using using `GetMethod("Hello", new[] { typeof(string) })` return a method with the signature `GetMethod(string?)`? (If it doesn't, you just lost the ability to add nullability annotations to the entire BCL because it'd break reflection uses. If it does, you break expectations.) Does it depend on whether your code was built as nullable? (If it does, how the heck does that interact with mixed-nullability projects/files? Now the compiler has to know about all the reflection methods or something.) You'd also be introducing this weird inconsistency where nullable value types and nullable reference types are represented differently. If you aren't going to surface it in reflection, why encode it in the type system at all? Finally nullability is try-state: Nullable, not-nullable, and oblivious. Do you surface oblivious too?

On a semi-related note: I originally thought they made `typeof(string?)` an error was because they wanted to reserve the possibility of making it special in the future, but this actually is seemingly not the case because `typeof(List<string?>)` *is* legal. It is mentioned in https://github.com/dotnet/roslyn/issues/29894#issuecomment-482274990 along with justification for the consistency, but it's not directly stated why `typeof(string?)` is illegal in the first place. (If I had to guess, it's probably just to make it immediately clear there's no special treatment of a nullable reference type in `typeof`.)

So in short: Having the nullability of a reference type represented in the type system would be a complex and confusing nightmare.

**Reply by skrysm** [reply]
Thanks for the elaborate answer. :) 

I was thinking of this in the same way as `int` and `int?` are two distinct types in the type system. So why should `string` and `string?` be the same?

But as you mentioned, doing this *now* (after there has been no distinction for years/decades) would probably break a lot of existing code. Especially since you can mix code *with* nullable reference types with code *without* nullable reference types.

**Reply by PathogenDavid** [reply]
No problem!

Another thing to keep in mind is that `int` and `int?` are different in the type system out of necessity. It isn't possible to truly have `int?` without keeping track of extra information.

We have a [convention going as far back as 1965](https://www.infoq.com/presentations/Null-References-The-Billion-Dollar-Mistake-Tony-Hoare/) where we reserved a specific value (generally 0) for pointers/reference types to be universally understood as `null`. No such _universal_ convention exists for arbitrary segments of data (whether that be a 32 bit `int` representing someone's age, a 64 bit `DateTime`, or something else), so you have to allocate an additional location to store whether or not the data is invalid. In the case of C# with .NET, we do this with the special wrapper type [`System.Nullable<T>`](https://github.com/dotnet/runtime/blob/5078dd61a81c88dbdd131a77b8148b4dee3803fa/src/libraries/System.Private.CoreLib/src/System/Nullable.cs).
